File: class_nplus1/puzzle.hs
-- Debugging
-- Searches over lists
-- Performance stuff

import Debug.Trace (trace)
import Control.Monad (guard)

add a b = trace (">>adding " ++ (show a) ++ " and " ++ (show b) ++ "<<") $ a + b

main1 = do
  xStr <- getLine
  yStr <- getLine
  let x = read xStr :: Int
  let y = read yStr :: Int
  let s = add x 12
  putStrLn $ show s


die = [1..6]

f = do
  x <- die
  y <- die
  z <- die
  guard $ x + y + z == 9
  [(x, y, z)]

g =
  die >>=
  (\x -> die >>=
    (\y -> die >>=
      (\z -> [(x,y,z)] >>
        (guard $ x+y+z == 9) >>
        [(x,y,z)]
      )
    )
  )

--   send 
-- + more
-- ------
--  money

digits = [0..9] :: [Int]

remove :: [Int] -> [Int] -> [Int]
remove [] bs = bs
remove (a:as) bs = filter (/= a) $ remove as bs

digitsToNum :: [Int] -> Int
digitsToNum ds = (go 0 ds)
  where
    go :: Int -> [Int] -> Int
    go acc (d:ds) = go (10*acc + d) ds
    go acc [] = acc

solution = do
  s <- remove [0] digits
  m <- remove [s, 0] digits

  let minSend = s*1000
  let maxSend = s*1000 + 999

  let minMore = m*1000
  let maxMore = m*1000 + 999

  let minMoney = m * 10000
  let maxMoney = m * 10000 + 9999

  guard $ minSend + minMore <= maxMoney
  guard $ maxSend + maxMore >= minMoney

  e <- remove [s, m] digits
  n <- remove [s, m, e] digits
  d <- remove [s, m, e, n] digits
  let send = digitsToNum [s,e,n,d]
  o <- remove [s, e, n, d, m] digits
  r <- remove [s, e, n, d, m, o] digits
  let more = digitsToNum [m, o, r, e]
  let sendPlusMore = send + more
  y <- remove [s, e, n, d, m, o, r] digits
  let money = digitsToNum [m, o, n, e, y]
  guard $ send + more == money
  [(send, more, money)]

main = putStrLn $ show solution

File: class_nplus1/puzzle.c
#include <stdio.h>

void solve()
{
  for (int s=1; s<10; s++) {
    for (int e=0; e<10; e++) {
      if (e == s) continue;
      for (int n=0; n<10; n++) {
        if (n == s || n == e) continue;
        for (int d=0; d<10; d++) {
          if (d == s || d == e || d == n) continue;
          int send = 1000*s + 100*e + 10*n + d;

          for (int m=1; m<10; m++) {
            if (m == s || m == e || m == n || m == d) continue;
            for (int o=0; o<10; o++) {
              if (o == s || o == e || o == n || o == d || o == m) continue;
              for (int r=0; r<10; r++) {
                if (r == s || r == e || r == n || r == d || r == m || r == o) continue;
                int more = 1000*m + 100*o + 10*r + e;

                for (int y=0; y<10; y++) {
                  if (y == s || y == e || y == n || y == d || y == m || y == o || y == r) continue;
                  int money = 10000*m + 1000*o + 100*n + 10*e + y;

                  if (send + more == money) {
                    printf("   %d%d%d%d\n", s, e, n, d);
                    printf("+  %d%d%d%d\n", m, o, r, e);
                    printf("-------\n");
                    printf("  %d%d%d%d%d\n", m, o, n, e, y);
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}

void main()
{
  solve();
}
